### PR TITLE
Add Globalization tests and expose more APIs in Globalization and String class

### PR DESCRIPTION
**Pull request by tarekgh**

### CODE REVIEW

**Comment by tarekgh** [comment]
CC @weshaggard @danmosemsft PTAL, thanks.

**Comment by tarekgh** [comment]
Forgot to mention the tests are enabled for Windows only for now and will enable it for other platforms when we support the functionality there.

File: src/System.Globalization/tests/CharUnicodeInfo/CharUnicodeInfoTests.netstandard1.7.cs [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+        {
+            for (int i=0; i<s_nonNumericsCodepoints.Length; i++)
+            {
+                Assert.Equal(CharUnicodeInfo.GetDecimalDigitValue(s_nonNumericsCodepoints[i]), -1);
+                Assert.Equal(CharUnicodeInfo.GetDecimalDigitValue(s_nonNumericsCodepoints, i), -1);
+            }
+        }
+
+        [Fact]
+        [PlatformSpecific(Xunit.PlatformID.Windows)] // Non Windows is not supported yet
```
Comment (danmosemsft) [inline src/System.Globalization/tests/CharUnicodeInfo/CharUnicodeInfoTests.netstandard1.7.cs]: >yet [](start = 85, length = 3)

is there an ID you can note for non Windows?

File: src/System.Globalization/tests/CharUnicodeInfo/CharUnicodeInfoTests.netstandard1.7.cs [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        public static void DigitsTest()
+        {
+            Assert.Equal(s_numericNonDecimalCodepoints.Length, s_numericNonDecimalValues.Length);
+            for (int i=0; i<s_numericNonDecimalCodepoints.Length; i++)
```
Comment (danmosemsft) [inline src/System.Globalization/tests/CharUnicodeInfo/CharUnicodeInfoTests.netstandard1.7.cs]: >i<s [](start = 26, length = 3)

nit: spacing .. is there an autoformatter?

File: src/System.Globalization/tests/CultureInfo/CultureInfoTests.netstandard1.7.cs [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        {
+            CultureInfo ci = CultureInfo.GetCultureInfo("ja-JP");
+            Assert.True((object) ci == (object) CultureInfo.GetCultureInfo("ja-JP"));
+            ci.ClearCachedData();
+            Assert.False((object) ci == (object) CultureInfo.GetCultureInfo("ja-JP"));
+        }
+
+        [Fact]
+        [PlatformSpecific(Xunit.PlatformID.Windows)] // Non Windows is not supported yet
+        public void CultureNotFoundExceptionTest()
```
Comment (danmosemsft) [inline src/System.Globalization/tests/CultureInfo/CultureInfoTests.netstandard1.7.cs]: >() [](start = 48, length = 2)

passing null and empty string?

File: src/System.Globalization/tests/System.Globalization.Tests.csproj [hunk 4]
```diff
@@ -152,6 +158,22 @@
       <Name>System.Globalization</Name>
     </ProjectReference>
   </ItemGroup>
+  <ItemGroup Condition="'$(TargetGroup)' == 'netstandard1.7'">
```
Comment (danmosemsft) [inline src/System.Globalization/tests/System.Globalization.Tests.csproj]: >ItemGroup [](start = 3, length = 9)

would it be clearer to move all these tests into system.runtime.tests, since the types are there now?

File: src/System.Globalization/tests/RegionInfo/RegionInfoTests.netstandard1.7.cs [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+            Assert.Equal(ri.CurrencyNativeName, currencyNativeName);
+            Assert.Equal(ri.ThreeLetterISORegionName, threeLetterISORegionName);
+            Assert.Equal(ri.ThreeLetterWindowsRegionName, threeLetterWindowsRegionName);
+        }
+
+        [Fact]
+        [PlatformSpecific(Xunit.PlatformID.Windows)] // Non Windows is not supported yet
+        public void NegativeTest()
+        {
+            Assert.Throws<System.ArgumentException>("name", () => new RegionInfo(""));
```
Comment (danmosemsft) [inline src/System.Globalization/tests/RegionInfo/RegionInfoTests.netstandard1.7.cs]: >("")); [](start = 80, length = 6)

RegoinInfo(null) ?

Comment (danmosemsft) [inline src/System.Globalization/tests/RegionInfo/RegionInfoTests.netstandard1.7.cs]: Oh, I see that is already in RegionInfoTests.Methods.cs

---
In reply to: [78286978](https://github.com/dotnet/corefx/pull/11594#discussion_r78286978) [](ancestors = 78286978)

File: src/System.Globalization/tests/CultureInfo/CultureInfoTests.netstandard1.7.cs (hunk 3 above)
Comment (danmosemsft) [inline src/System.Globalization/tests/CultureInfo/CultureInfoTests.netstandard1.7.cs]: never mind, already in existing Ctor_String_Invalid()

---
In reply to: [78286935](https://github.com/dotnet/corefx/pull/11594#discussion_r78286935) [](ancestors = 78286935)

Comment (danmosemsft) [inline src/System.Globalization/tests/CultureInfo/CultureInfoTests.netstandard1.7.cs]: this test could just go in there.

---
In reply to: [78287001](https://github.com/dotnet/corefx/pull/11594#discussion_r78287001) [](ancestors = 78287001,78286935)

**Comment by danmosemsft** [comment]
Thorough tests Tarek!

**Comment by danmosemsft** [comment]
:shipit:

File: src/System.Globalization.Calendars/tests/Misc/Calendars.netstandard1.7.cs [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+// See the LICENSE file in the project root for more information.
+
+using Xunit;
+using System.Globalization;
+using System.Collections.Generic;
+
+namespace System.Globalization.Tests
+{
+
+    public static class CalendarsMoreTests
```
Comment (stephentoub) [inline src/System.Globalization.Calendars/tests/Misc/Calendars.netstandard1.7.cs]: Calendars_netstandard17 if that's the defining characteristic of why these tests are here?

File: src/System.Globalization.Calendars/tests/Misc/Calendars.netstandard1.7.cs [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+            yield return new object[] { new PersianCalendar()           , 0     , CalendarAlgorithmType.SolarCalendar       };
+            yield return new object[] { new TaiwanCalendar()            , 0     , CalendarAlgorithmType.SolarCalendar       };
+            yield return new object[] { new TaiwanLunisolarCalendar()   , 0     , CalendarAlgorithmType.LunisolarCalendar   };
+            yield return new object[] { new ThaiBuddhistCalendar()      , 0     , CalendarAlgorithmType.SolarCalendar       };
+            yield return new object[] { new UmAlQuraCalendar()          , 0     , CalendarAlgorithmType.LunarCalendar       };
+        }
+
+        [Theory]
+        [MemberData(nameof(Calendars_TestData))]
+        [PlatformSpecific(Xunit.PlatformID.Windows)] // Non Windows is not supported yet
```
Comment (stephentoub) [inline src/System.Globalization.Calendars/tests/Misc/Calendars.netstandard1.7.cs]: But unix will be supported?  Might be better then to use:
```C#
[ActiveIssue(issueNumberForEnablingUnix, Xunit.PlatformID.AnyUnix)]
```
for this and others like it.

File: src/System.Globalization.Calendars/tests/Misc/Calendars.netstandard1.7.cs [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+        public static void CloningTest(Calendar calendar, int yearHasLeapMonth, CalendarAlgorithmType algorithmType)
+        {
+            Calendar cloned = (Calendar) calendar.Clone();
+            Assert.Equal(cloned.GetType(), calendar.GetType());
```
Comment (stephentoub) [inline src/System.Globalization.Calendars/tests/Misc/Calendars.netstandard1.7.cs]: Nit: the first parameter to Equal is "expected", the second is "actual", so these args should be reversed, e.g.
```C#
Assert.Equal(calendar.GetType(), cloned.GetType());
```
This matters primarily because it affects the error message if the test fails, and it can be confusing if they're swapped.

File: src/System.Globalization.Calendars/tests/Misc/Calendars.netstandard1.7.cs [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+
+        [Theory]
+        [MemberData(nameof(Calendars_TestData))]
+        [PlatformSpecific(Xunit.PlatformID.Windows)] // Non Windows is not supported yet
+        public static void GetLeapMonthTest(Calendar calendar, int yearHasLeapMonth, CalendarAlgorithmType algorithmType)
+        {
+            if (yearHasLeapMonth > 0)
+            {
+                Assert.NotEqual(calendar.GetLeapMonth(yearHasLeapMonth),  0);
+                Assert.Equal(calendar.GetLeapMonth(yearHasLeapMonth - 1), 0);
```
Comment (stephentoub) [inline src/System.Globalization.Calendars/tests/Misc/Calendars.netstandard1.7.cs]: Same expected/actual comment as above.  I'll stop commenting on this, but it applies for all such asserts.

File: src/System.Globalization.Calendars/tests/Misc/Calendars.netstandard1.7.cs [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+        [Theory]
+        [MemberData(nameof(Calendars_TestData))]
+        [PlatformSpecific(Xunit.PlatformID.Windows)] // Non Windows is not supported yet
+        public static void ReadOnlyTest(Calendar calendar, int yearHasLeapMonth, CalendarAlgorithmType algorithmType)
+        {
+            Assert.False(calendar.IsReadOnly);
+            var readOnlyCal = Calendar.ReadOnly(calendar);
+            Assert.True(readOnlyCal.IsReadOnly);
+            var colnedCal = (Calendar) readOnlyCal.Clone();
+            Assert.False(colnedCal.IsReadOnly);
```
Comment (stephentoub) [inline src/System.Globalization.Calendars/tests/Misc/Calendars.netstandard1.7.cs]: Especially on unix where we don't currently get line numbers in stack traces, it's helpful in Assert.True/False calls to include a message that helps to differentiate them when there are multiple such asserts in the same method.  Otherwise, when one fails, we don't know which.

File: src/System.Globalization/tests/CharUnicodeInfo/CharUnicodeInfoTests.netstandard1.7.cs (hunk 1 above)
Comment (stephentoub) [inline src/System.Globalization/tests/CharUnicodeInfo/CharUnicodeInfoTests.netstandard1.7.cs]: Yeah, as noted above, I'd suggest we switch these to be ActiveIssue.  That helps both by including the relevant issue number and by highlighting that it's disabled temporarily rather than permanently on that platform(s).

File: src/System.Globalization/tests/CultureInfo/CultureInfoTests.netstandard1.7.cs [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+        {
+            Assert.Throws<CultureNotFoundException>("name", () => new CultureInfo("!@#$%^&*()"));
+            Assert.Throws<CultureNotFoundException>("name", () => new CultureInfo("This is invalid culture"));
+            Assert.Throws<CultureNotFoundException>("name", () => new CultureInfo("longCulture" + new string('a', 100)));
+            Assert.Throws<CultureNotFoundException>("culture", () => new CultureInfo(0x1000));
+
+            try { new CultureInfo("This is invalid culture"); }
+            catch (CultureNotFoundException e) { Assert.Equal("This is invalid culture", e.InvalidCultureName); }
```
Comment (stephentoub) [inline src/System.Globalization/tests/CultureInfo/CultureInfoTests.netstandard1.7.cs]: Assert.Throws returned thrown exception, so instead of doing this try/catch, you can just augment the case of this above:
```C#
CultureNotFoundException e = Assert.Throws<CultureNotFoundException>("name", () => new CultureInfo("This is invalid culture"));
Assert.Equal("This is invalid culture", e.InvalidCultureName);
```

**Comment by stephentoub** [comment]
Some comments, but generally LGTM.  Thanks for adding all of these tests, @tarekgh..

File: src/System.Globalization.Calendars/tests/Misc/Calendars.netstandard1.7.cs (hunk 6 above)
Comment (tarekgh) [inline src/System.Globalization.Calendars/tests/Misc/Calendars.netstandard1.7.cs]: The new tests mainly for the added APIs in net standard 1.7 and we are testing misc functionality of the calendars this is why I chose ti put the tests here.

File: src/System.Globalization.Calendars/tests/Misc/Calendars.netstandard1.7.cs (hunk 7 above)
Comment (tarekgh) [inline src/System.Globalization.Calendars/tests/Misc/Calendars.netstandard1.7.cs]: I'll do that.

File: src/System.Globalization.Calendars/tests/Misc/Calendars.netstandard1.7.cs (hunk 8 above)
Comment (tarekgh) [inline src/System.Globalization.Calendars/tests/Misc/Calendars.netstandard1.7.cs]: I'll fix that

File: src/System.Globalization.Calendars/tests/Misc/Calendars.netstandard1.7.cs (hunk 9 above)
Comment (tarekgh) [inline src/System.Globalization.Calendars/tests/Misc/Calendars.netstandard1.7.cs]: thanks, I'll take with that.

File: src/System.Globalization.Calendars/tests/Misc/Calendars.netstandard1.7.cs (hunk 10 above)
Comment (tarekgh) [inline src/System.Globalization.Calendars/tests/Misc/Calendars.netstandard1.7.cs]: I'll convert these into .Assert.Equal instead of adding messages.

Comment (stephentoub) [inline src/System.Globalization.Calendars/tests/Misc/Calendars.netstandard1.7.cs]: How will that help?

Comment (tarekgh) [inline src/System.Globalization.Calendars/tests/Misc/Calendars.netstandard1.7.cs]: sorry I mis-understood your comment. I'll keep the .True/.False and will add the message

File: src/System.Globalization/tests/CharUnicodeInfo/CharUnicodeInfoTests.netstandard1.7.cs (hunk 1 above)
Comment (tarekgh) [inline src/System.Globalization/tests/CharUnicodeInfo/CharUnicodeInfoTests.netstandard1.7.cs]: I am applying Stephen's suggestion.

File: src/System.Globalization/tests/CharUnicodeInfo/CharUnicodeInfoTests.netstandard1.7.cs (hunk 2 above)
Comment (tarekgh) [inline src/System.Globalization/tests/CharUnicodeInfo/CharUnicodeInfoTests.netstandard1.7.cs]: I'll fix that.

File: src/System.Globalization/tests/CultureInfo/CultureInfoTests.netstandard1.7.cs (hunk 11 above)
Comment (tarekgh) [inline src/System.Globalization/tests/CultureInfo/CultureInfoTests.netstandard1.7.cs]: I'll fix that.

File: src/System.Globalization/tests/System.Globalization.Tests.csproj (hunk 4 above)
Comment (tarekgh) [inline src/System.Globalization/tests/System.Globalization.Tests.csproj]: this will require moving the whole tests and not only the tests I am adding. I would say if we decide to do that we can do it in different PR. but IMHO I would say keeping the tests here is logical too as it is testing old contracts too in addition to the new one.

**Comment by tarekgh** [comment]
Test Innerloop CentOS7.1 Release Build and Test please

**Comment by tarekgh** [comment]
Thanks @stephentoub and @danmosemsft for your review.

File: src/System.Globalization.Calendars/tests/System.Globalization.Calendars.Tests.csproj [hunk 12]
```diff
@@ -8,8 +8,14 @@
     <OutputType>Library</OutputType>
     <AssemblyName>System.Globalization.Calendars.Tests</AssemblyName>
     <RootNamespace>System.Globalization.Calendars.Tests</RootNamespace>
-    <NugetTargetMoniker>.NETStandard,Version=v1.3</NugetTargetMoniker>
+    <NugetTargetMoniker Condition="'$(TargetGroup)'=='netstandard1.7' AND '$(NugetTargetMoniker)'==''">.NETStandard,Version=v1.7</NugetTargetMoniker>
```
Comment (weshaggard) [inline src/System.Globalization.Calendars/tests/System.Globalization.Calendars.Tests.csproj]: nit: This property is redundant and can be removed.

File: src/System.Globalization/tests/System.Globalization.Tests.csproj (hunk 4 above)
Comment (weshaggard) [inline src/System.Globalization/tests/System.Globalization.Tests.csproj]: I agree we should probably keep these here until we decide we don't care about testing the older contract versions.
